airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py

```diff
@@ -90,27 +124,46 @@ def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Ite
             field=self.cursor_field,
             days_of_data_storage=self.days_of_data_storage,
             range_days=self.range_days,
+            time_zone=self.time_zone,
         )
 
-    def get_date_params(self, stream_slice: Mapping[str, Any], cursor_field: str, end_date: pendulum.datetime = None):
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
         """
-        Returns `start_date` and `end_date` for the given stream_slice.
-        If (end_date - start_date) is a big date range (>= 1 month), it can take more than 2 hours to process all the records from the given slice.
-        After 2 hours next page tokens will be expired, finally resulting in page token expired error
-        Currently this method returns `start_date` and `end_date` with 15 days difference.
+        This method is overridden to handle GoogleAdsException with EXPIRED_PAGE_TOKEN error code,
+        and update `start_date` key in the `stream_slice` with the latest read record's cursor value, then retry the sync.
         """
-
-        end_date = end_date or pendulum.yesterday(tz=self.time_zone)
-        start_date = pendulum.parse(stream_slice.get(cursor_field))
-        if start_date > pendulum.now():
-            return start_date.to_date_string(), start_date.add(days=1).to_date_string()
-
-        end_date = min(end_date, pendulum.parse(stream_slice.get(cursor_field)).add(days=self.range_days))
-
-        # Fix issue #4806, start date should always be lower than end date.
-        if start_date.add(days=1).date() >= end_date.date():
-            return start_date.add(days=1).to_date_string(), start_date.add(days=2).to_date_string()
-        return start_date.add(days=1).to_date_string(), end_date.to_date_string()
+        state = stream_state or {}
+        while True:
+            try:
+                response = self.google_ads_client.send_request(self.get_query(stream_slice))
+                for record in self.parse_response(response):
+                    state = self.get_updated_state(state, record)
```

Comment:
not sure I understand why we need to call `get_updated_state`  here

Reply:
@keu This method is called to get the cursor value from the record. To be sure the cursor value is the latest state I call this method.